File: src/scrapers/sacnilk.ts
import * as fs from 'fs';
import * as path from 'path';

const BASE_URL = "https://www.sacnilk.com";
const MAIN_URL = `${BASE_URL}/metasection/box_office`;
const DATA_DIR = path.resolve(__dirname, '../../data');
const OUTPUT_FILE = path.join(DATA_DIR, "sacnilk_data.json");

if (!fs.existsSync(DATA_DIR)) {
  fs.mkdirSync(DATA_DIR, { recursive: true });
}

function normalizeTitle(title: string) {
  const cleaned = title.replace(/\b(19|20)\d{2}\b/, "");
  return cleaned.toLowerCase().replace(/[^a-z0-9]+/gi, " ").trim();
}

function cleanMovieTitle(title: string) {
  return title.replace(/\s+Box Office.*$/i, "").trim();
}

async function fetchHTML(url: string) {
  const res = await fetch(url, {
    headers: { "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36" }
  });
  if (!res.ok) throw new Error(`Failed to fetch ${url}`);
  return await res.text();
}

async function extractMovieLinks() {
  console.log("🌐 Fetching SACNilk Main Page...");
  const html = await fetchHTML(MAIN_URL);

  const movieMap: any = {};
  const regex = /<a\s+href="(?:https:\/\/www\.sacnilk\.com)?(\/news\/[^"]+Box_Office_Collection[^"]*)"/gi;

  let match;
  while ((match = regex.exec(html))) {
    let href = match[1];
    
    // Extract movie name from URL: e.g. /news/Immortal_2026_Box_Office_Collection_Day_Wise_Worldwide
    const urlParts = href.split('/news/')[1].split('_Box_Office_Collection');
    if (urlParts.length > 0) {
      let rawTitle = urlParts[0].replace(/_/g, ' ').replace(/\b(19|20)\d{2}\b/, '').trim();
      
      const normalized = normalizeTitle(rawTitle);

      if (!movieMap[normalized]) movieMap[normalized] = [];
      movieMap[normalized].push({
        name: rawTitle,
        link: BASE_URL + href,
        day: 1 // Default to 1, we will scrape the actual day from the page
      });
    }
  }

  const finalMovies: any[] = [];
  Object.values(movieMap).forEach((entries: any) => {
    entries.sort((a: any, b: any) => b.day - a.day); // Get latest day
    if (entries[0].day <= 30) {
      finalMovies.push(entries[0]);
    }
  });

  return finalMovies;
}

async function extractAmountCr(url: string) {
  try {
    const html = await fetchHTML(url);
    
    // Look for: Overall Total Collection (Net)... ₹<amount>Cr
    const match = html.match(/Overall Total Collection\s*\(Net\)[\s\S]*?₹([\d.]+)\s*Cr/i);
    
    if (match) {
      return parseFloat(match[1]);
    }
    
    // Fallback if not found
    console.log("Could not find 'Overall Total Collection (Net)' for", url);
  } catch (e: any) {
    console.error("❌ Error fetching amount from:", url, e.message);
  }
  return null;
}

async function runScraper() {
  console.log("🚀 Starting SACNilk Industry Scraper...");
  
  try {
    const movies = await extractMovieLinks();
    console.log(`🎬 Found ${movies.length} active movies on SACNilk.`);
    
    const results = [];
    
    for (const m of movies) {
      console.log(`fetching day ${m.day} for ${m.name}...`);
      const amount = await extractAmountCr(m.link);
      
      results.push({
        movie: m.name,
        day: m.day,
        estimateCr: amount || 0,
        sourceUrl: m.link,
        lastUpdated: new Date().toISOString()
      });
      
      // Gentle delay
      await new Promise(r => setTimeout(r, 1000));
    }
    
    fs.writeFileSync(OUTPUT_FILE, JSON.stringify(results, null, 2));
    console.log(`✅ Saved SACNilk industry estimates to ${OUTPUT_FILE}`);
  } catch (err) {
    console.error("Failed to scrape SACNilk:", err);
  }
}

runScraper();

File: src/scrapers/sacnilk.py
import json
import os
import datetime
import time
from pathlib import Path
import requests

DATA_DIR = Path(__file__).parent.parent.parent / "data"
os.makedirs(DATA_DIR, exist_ok=True)
OUTPUT_FILE = DATA_DIR / "sacnilk_data.json"

API_URL = "https://sacapi.text2024mail.workers.dev/?date="

def fetch_sacnilk(date_str, retries=3):
    url = f"{API_URL}{date_str}"
    
    headers = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"}
    for attempt in range(retries):
        try:
            response = requests.get(url, headers=headers, timeout=15)
            if response.status_code == 200:
                return response.json()
            else:
                print(f"[{date_str}] HTTP {response.status_code}, retry...")
        except Exception as e:
            print(f"[{date_str}] Error: {e}")
        
        time.sleep(2 ** attempt)
    return []

def main():
    print("🚀 Starting Sacnilk Industry Estimates Scraper...")
    
    today = datetime.datetime.now(datetime.timezone(datetime.timedelta(hours=5, minutes=30))).date()
    live_date_str = today.strftime("%Y-%m-%d")
    
    print(f"📡 Fetching Live Industry Estimates for {live_date_str}...")
    data = fetch_sacnilk(live_date_str)
    
    if data:
        with open(OUTPUT_FILE, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
        print(f"✅ Saved Sacnilk estimates for {len(data)} items.")
    else:
        print("⚠️ No Sacnilk data returned.")
        if not OUTPUT_FILE.exists():
            with open(OUTPUT_FILE, "w", encoding="utf-8") as f:
                json.dump([], f)

if __name__ == "__main__":
    main()
